### PR TITLE
Add rest api helper interface to access non-xcat resource

### PR DESCRIPTION
**Pull request by chenglch**
Currently xcat rest service is running with apache user who do
not have the privilege to manage some local resource like credential
files needed by xcat web ui to get the conserver access.

This patch add scripts in xcatconfig to generate credential files
for conserver user. This patch also add a work around framework for
xcat restapi to provide the interface to get or set the system resource
so that restapi with apache privilege can retrieve the credential files.

### CODE REVIEW

**Comment by daniceexi** [comment]
@zet809 Could you review this pull request?

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+#-------------------------------------------------------
+sub process_request
+{
+    my $request = shift;
+    $::callback = shift;
+    my $subreq  = shift;
+    my $command = $request->{command}->[0];
+
+    if ($command eq "localrest") {
+        return handle_rest_request($request, $subreq);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: The subroutine declared in line 74 need 3 parametes, but only 2 passed here.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    if (scalar(@args) < 2) {
+        $rsp->{data}->[0] = "Local rest api take at least two parameter.";
+        xCAT::MsgUtils->message("E", $rsp, $::callback);
+        return 1;
+    }
+    $method     = shift @args;
+    $resource   = shift @args;
+    $subroutine = $method . '_' . $resource;
+    @params     = @args;
+    my @subroutines = $list_module_func->("xCAT_plugin::localrest");
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Using sentence like "if (!defined(&$subroutine)) {...}" maybe easier.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    my $result = $subroutine->(@params);
+
+    # handle the result from the rest subroutine
+    if (ref($result) eq 'HASH') {
+        if (defined($result->{'type'}) && $result->{'type'} eq 'stream'
+            && defined($result->{'filename'})) {
+            $rsp->{data}->[0] = "stream";
+            $rsp->{data}->[1] = $result->{'filename'};
+            $rsp->{data}->[2] = $result->{'data'};
+        } else {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: When will this case happen?

Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: For most case , we do not need to change this subroutine, in a normal situation,  a request withHASH return will match this `else` condition to return json message generated from a hash object like ('state' => 'on') -> {  'state': 'on' } 
 The `stream` type above is for downloading.

File: xCAT-server/xCAT-wsapi/xcatws.cgi [hunk 4]
```diff
@@ -1230,6 +1230,24 @@ my %URIdef = (
...
+    ### localres can be looked as a top level non-xcat resource pool
+    ### rest operation will trasfer the target resource to xcatd plugin to handle the non-xcat resource
+    localres => {
+        localres => {
+            desc => "[URI:/localres/*] - The local non-xcat resource.",
+            matcher => '^/localres(/[^/]*)+$',
+            GET => {
+                desc => "List information for the target system resource.",
+                usage => "||For target resource can match any resource type which can be proccssed by the resthelper plugin.|",
+                example => qq(|List adapters on MN machine|GET|/localres/interface/|{\n   \"interfaces\":[\n\"eth0\",\n         \"eht1\",\n]"|),
```
Comment (zet809) [inline xCAT-server/xCAT-wsapi/xcatws.cgi]: eht1==>eth1

**Comment by zet809** [comment]
Looks good to me.

File: xCAT-server/xCAT-wsapi/xcatws.cgi (hunk 4 above)
Comment (chenglch) [inline xCAT-server/xCAT-wsapi/xcatws.cgi]: ?
this is an array example
"interfaces": ["eth0", "eth1"]

**Comment by zet809** [comment]
Looks good to me, I agree to merge it.

File: xCAT-server/sbin/xcatconfig [hunk 5]
```diff
@@ -448,12 +448,6 @@ if ($::INITIALINSTALL || $::FORCE || $::UPDATEINSTALL || $::genCredentials)
     }
     else
     {
-        #since the xcatd service unit file is updated on xcat upgrade
```
Comment (immarvin) [inline xCAT-server/sbin/xcatconfig]: hi @chenglch , why these lines are removed?

**Comment by chenglch** [comment]
@immarvin , sorry this is my fault to deal with the conflict when I overwrite this file.

**Comment by immarvin** [comment]
hi @chenglch , would you please create a new PR and recover these lines? thanks

**Comment by chenglch** [comment]
@immarvin , I have create a new PR #1588 to fix this conflict error

**Pull request merged**